adcc/Intermediates.py

```diff
@@ -0,0 +1,75 @@
+#!/usr/bin/env python3
+## vi: tabstop=4 shiftwidth=4 softtabstop=4 expandtab
+## ---------------------------------------------------------------------
+##
+## Copyright (C) 2020 by the adcc authors
+##
+## This file is part of adcc.
+##
+## adcc is free software: you can redistribute it and/or modify
+## it under the terms of the GNU General Public License as published
+## by the Free Software Foundation, either version 3 of the License, or
+## (at your option) any later version.
+##
+## adcc is distributed in the hope that it will be useful,
+## but WITHOUT ANY WARRANTY; without even the implied warranty of
+## MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+## GNU General Public License for more details.
+##
+## You should have received a copy of the GNU General Public License
+## along with adcc. If not, see <http://www.gnu.org/licenses/>.
+##
+## ---------------------------------------------------------------------
+from .timings import Timer
+from .functions import evaluate
+
+
+class Intermediates():
+    """
+    Class offering to return a number of intermediate tensors.
+    The tensors are cached internally for later reuse
+    """
+    generators = {}      # Registered generator functions
+
+    def __init__(self, ground_state):
+        self.ground_state = ground_state
+        self.reference_state = ground_state.reference_state
+        self.timer = Timer()
+        self.cached_tensors = {}  # Cached tensors
+        # TODO Make caching configurable ??
+
+    def __getattr__(self, key):
+        if key in self.cached_tensors:
+            return self.cached_tensors[key]
+        elif key in self.generators:
+            # Evaluate the tensor, all generators take (hf, mp, intermediates)
+            generator = self.generators[key]
+            with self.timer.record(key):
+                tensor = generator(self.reference_state, self.ground_state, self)
+                self.cached_tensors[key] = evaluate(tensor)
+            return self.cached_tensors[key]
+        else:
+            raise AttributeError
+
+    def clear(self):
+        """Clear all cached tensors to free storage"""
+        self.cached_tensors.clear()
+
+    def __repr__(self):
+        return (
+            "AdcIntermediates(contains="
+            + list(self.cached_tensors.keys()).join(",")
+            + ")"
+        )
+
+
+def register_as_intermediate(function):
```

Comment:
This is nice. Can/should we somehow reuse this also when we move `mp` py-side?

Reply:
Maybe. Not sure yet. I think MP is a bit different as the tradeoffs are different since it is the ground state it should always be cached, I suppose. The think is that `@register_as_intermediate` allows to keep the implementation of the working equation and the implementation of the intermediates close, such that you can easily understand the whole lot at once. For LazyMp I don't see why this extra wrapper would be needed. One could just use an `@cached_property` for that or something similar.